lifecycle.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+)
+
+// Lifecycle is a representation of a large process that needs to be executed, like building and deploying a sample to
+// Cloud Run. Each lifecycle is made up of a list of different phases -- like build, deploy, and post-deploy -- and each
+// phase is in turn made up of small goals that act as the steps of each phase.
+type Lifecycle struct {
+	// id is a unique identifier for this Lifecycle. Not currently used.
+	id string
+
+	// phases contains an order list of phases associated with this lifecycle.
+	phases []*Phase
+
+	// phaseMap is a map that maps IDs to their associated phase in the Phases slice.
+	phaseMap map[string]*Phase
+}
+
+// Phase represents one phase of a larger lifecycle. It has a list of goals that make up the actual step-by-step
+// execution of the phase in the form of exec.Cmd. In addition, each phase has a string ID that will uniquely identify
+// it in the context of its parent Lifecycle.
+type Phase struct {
+	id    string
+	goals []*exec.Cmd
+}
+
+// newLifeCycle builds a new Lifecycle given the provided phases and lifecycle id.
+func newLifeCycle(id string, phases []*Phase) *Lifecycle {
+	lifecycle := Lifecycle{
+		id:     id,
+		phases: phases,
+	}
+
+	lifecycle.phaseMap = make(map[string]*Phase)
+	for _, phase := range phases {
```

Comment:
https://github.com/golang/go/wiki/CodeReviewComments#variable-names
```suggestion
	for _, p := range phases {
```